File: website/lib/service-area.ts
/**
 * Hizmet alanı — nereye teslimat yapıyoruz?
 *
 * Faz 1'de yalnızca **Konya / Selçuklu** ve **Konya / Karatay**. Müşteri il ve
 * ilçeyi serbest yazamaz; formda il sabit, ilçe listeden gelir.
 *
 * Bu dosya Dart tarafındaki `packages/core/lib/src/service_area.dart` ve sunucu
 * tarafındaki `ServiceArea.php` ile **aynı değerleri** taşır. Üçü birlikte
 * değişir; biri unutulursa istemcinin kabul ettiği adresi sunucu reddeder.
 */

export const SERVICE_AREA_CITY = 'Konya';

/** Teslimat yapılan ilçeler; sıra formdaki sıradır. */
export const SERVICE_AREA_DISTRICTS = ['Selçuklu', 'Karatay'] as const;

export type ServiceAreaDistrict = (typeof SERVICE_AREA_DISTRICTS)[number];

/**
 * Türkçe'ye duyarlı küçük harf.
 *
 * `toLowerCase()` dilden bağımsızdır ve `I` harfini `i`'ye düşürür; Türkçe'de
 * doğrusu `ı`'dır. Bugünkü iki ilçe adında fark yaratmıyor ama karşılaştırmayı
 * baştan doğru yazmak, ileride eklenecek bir ilçede sessiz bir eşleşmemeden
 * ucuz.
 */
function trLower(value: string): string {
  return value.replace(/I/g, 'ı').replace(/İ/g, 'i').toLowerCase();
}

export function coversDistrict(value: string | null | undefined): boolean {
  if (!value) return false;
  const needle = trLower(value.trim());
  return SERVICE_AREA_DISTRICTS.some((district) => trLower(district) === needle);
}

export function coversCity(value: string | null | undefined): boolean {
  if (!value) return false;
  return trLower(value.trim()) === trLower(SERVICE_AREA_CITY);
}

/**
 * Haritanın hapsedildiği kutu — `packages/core/lib/src/service_area.dart`
 * ile AYNI değerler (W-16).
 *
 * Üç yerde birden duruyor (Dart, PHP, buradaki TypeScript) ve birlikte
 * değişmek zorundalar; biri unutulursa istemcinin kabul ettiği iğneyi
 * sunucu reddeder.
 *
 * Kenarlar DAHİL: tam sınırdaki bir iğneyi reddetmek, kutuyu bir santimetre
 * içeriden çizmekle aynı şey olurdu.
 */
export const SERVICE_AREA_BOUNDS = {
  south: 37.8,
  north: 38.1,
  west: 32.35,
  east: 32.75,
} as const;

/** Harita ilk açıldığında ortalanacak nokta (Konya merkez). */
export const SERVICE_AREA_CENTER = { latitude: 37.8746, longitude: 32.4932 } as const;

/**
 * Kutunun tamamını gösteren en uzak seviye. Daha uzağa izin verilirse
 * "haritayı kutuya hapset" kısıtı sağlanamaz ve harita donmuş gibi davranır.
 */
export const SERVICE_AREA_MIN_ZOOM = 12.5;

/** Karo sağlayıcısının (OSM) verdiği en yakın seviye. */
export const SERVICE_AREA_MAX_ZOOM = 19;

/** Nokta hizmet alanı kutusunun içinde mi? */
export function containsPoint(latitude: number, longitude: number): boolean {
  return (
    latitude >= SERVICE_AREA_BOUNDS.south &&
    latitude <= SERVICE_AREA_BOUNDS.north &&
    longitude >= SERVICE_AREA_BOUNDS.west &&
    longitude <= SERVICE_AREA_BOUNDS.east
  );
}

File: packages/core/lib/src/service_area.dart
/// Hizmet alanı — nereye teslimat yapıyoruz?
///
/// Faz 1'de yalnızca **Konya / Selçuklu** ve **Konya / Karatay**. Müşteri il
/// ve ilçeyi serbest yazamaz; formlarda ikisi de kilitlidir ve harita bu
/// kutunun dışına çıkmaz.
///
/// ## Neden burada
///
/// Kural üç istemcide de aynı görünmeli: mobil form, web formu ve haritanın
/// sınırı. Dart tarafının tek kaynağı burasıdır. Web sitesi (TypeScript) ve
/// sunucu (PHP) aynı değerleri kendi dillerinde tekrarlar — karşılıkları
/// `website/lib/service-area.ts` ve
/// `platform/extensions/veykemtu/bridgeapi/src/Services/ServiceArea.php`.
/// Üçü DEĞİŞİRSE BİRLİKTE değişir; biri unutulursa istemci kabul ettiği
/// adresi sunucuya kabul ettiremez.
///
/// Bölge sayısı arttığında bu liste sunucudan gelmeli (`docs/11-yol-haritasi.md`
/// §bölgeler); üç ilçe için üç yerde sabit tutmak, sözleşmeye alan eklemekten
/// ucuzdur.
library;

import 'escpos/turkish_case.dart';

/// Teslimat yapılan il ve ilçeler.
abstract final class ServiceArea {
  /// Tek il. Formda seçilebilir değil, sabit gösterilir.
  static const String city = 'Konya';

  /// Teslimat yapılan ilçeler. Sıra formdaki sıradır.
  static const List<String> districts = <String>['Selçuklu', 'Karatay'];

  /// Kutunun güney kenarı (enlem).
  ///
  /// ## Kutu neden ilçe sınırı değil
  ///
  /// Değerler iki ilçenin **yerleşik alanını** kapsayan bir dikdörtgendir,
  /// idari sınır poligonu değil. Kenarlarda komşu ilçelerden (özellikle
  /// güneybatıda Meram) bir miktar alan da kutuya girer. Poligon denetimi
  /// sınır verisi gerektirir ve bir dikdörtgenin çözdüğü asıl sorunu —
  /// müşterinin haritayı Ankara'ya kaydırıp orayı işaretlemesi — zaten
  /// çözülmüş sayıyoruz.
  static const double south = 37.80;

  /// Kuzey kenarı — Selçuklu'nun kuzey yerleşimlerini içerir.
  static const double north = 38.10;

  /// Batı kenarı.
  static const double west = 32.35;

  /// Doğu kenarı — Karatay'ın doğu yerleşimlerini içerir.
  static const double east = 32.75;

  /// Harita iğnesiz açıldığında bakılan yer: Konya merkez.
  static const double centerLatitude = 37.8746;
  static const double centerLongitude = 32.4932;

  /// Kutu bu yakınlıktan daha uzağa çekilemez.
  ///
  /// Daha uzağı, ekrana kutudan büyük bir alan sığdırırdı; o noktada
  /// "haritayı kutuya hapset" kısıtı sağlanamaz hâle gelir ve harita
  /// donmuş gibi davranır.
  static const double minZoom = 12.5;

  /// Karo sağlayıcısının (OSM) verdiği en yakın seviye.
  static const double maxZoom = 19;

  /// Verilen ilçe hizmet alanında mı?
  ///
  /// Karşılaştırma Türkçe'ye duyarlı küçük harfe indirilerek yapılır, çünkü
  /// eski kayıtlarda ilçe adı büyük harfle de geçebiliyor. Dart'ın dilden
  /// bağımsız `toLowerCase()`'i `I` harfini `i`'ye düşürür; Türkçe'de doğrusu
  /// `ı`'dır ve bu fark ileride eklenecek bir ilçede ("KADINHANI") sessiz bir
  /// eşleşmeme yaratır.
  static bool coversDistrict(String? district) {
    if (district == null) return false;
    final needle = TurkishCase.toLowerCase(district.trim());
    return districts.any((known) => TurkishCase.toLowerCase(known) == needle);
  }

  /// Verilen il hizmet alanında mı?
  static bool coversCity(String? value) {
    if (value == null) return false;
    return TurkishCase.toLowerCase(value.trim()) ==
        TurkishCase.toLowerCase(city);
  }

  /// Nokta kutunun içinde mi?
  ///
  /// Kenarlar dahildir: tam sınırdaki bir iğneyi reddetmek, kutuyu bir
  /// santimetre içeriden çizmekle aynı şey olurdu.
  static bool containsPoint(double latitude, double longitude) =>
      latitude >= south &&
      latitude <= north &&
      longitude >= west &&
      longitude <= east;
}
